core/src/main/scala/api/sfs.scala

```diff
@@ -25,15 +30,19 @@ trait Filesystem {
    *  things too eagerly, nor caching data for too long.
    */
   def resolve(path: Path): Key
-  def metadata(key: Key): Metadata
-  def lookup(key: Key): Data
+  def metadata(key: Key): M[Metadata]
+  def lookup(key: Key): M[Data]
```

Comment:
While I agree with your reasoning, I envision file systems based on content located elsewhere (database, network, ...). Over the last few years the world has embraced 'non-blocking', this is not without reason. Even the most basic metadata could come in asynchronously. Are you proposing we block on that?